# Learning Pandas here

In [44]:
import pandas as pd

## Loading CSV data

Pandas gives a great way to consume data from different format like CSV, JSON etc.

In [45]:
df = pd.read_csv('../../data/stackoverflow/survey_results_public.csv', index_col='ResponseId')
df.head()

Q120                      MainBranch              Age  \
ResponseId                                                             
1           I agree                   None of these  18-24 years old   
2           I agree  I am a developer by profession  25-34 years old   
3           I agree  I am a developer by profession  45-54 years old   
4           I agree  I am a developer by profession  25-34 years old   
5           I agree  I am a developer by profession  25-34 years old   

                                                   Employment  \
ResponseId                                                      
1                                                         NaN   
2                                         Employed, full-time   
3                                         Employed, full-time   
4                                         Employed, full-time   
5           Employed, full-time;Independent contractor, fr...   

                                      RemoteWork  \
ResponseId                                         
1                                            NaN   
2                                         Remote   
3           Hybrid (some remote, some in-person)   
4           Hybrid (some remote, some in-person)   
5                                         Remote   

                                             CodingActivities  \
ResponseId                                                      
1                                                         NaN   
2           Hobby;Contribute to open-source projects;Boots...   
3           Hobby;Professional development or self-paced l...   
4                                                       Hobby   
5           Hobby;Contribute to open-source projects;Profe...   

                                                 EdLevel  \
ResponseId                                                 
1                                                    NaN   
2           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
5           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                                    LearnCode  \
ResponseId                                                      
1                                                         NaN   
2           Books / Physical media;Colleague;Friend or fam...   
3           Books / Physical media;Colleague;On the job tr...   
4           Colleague;Friend or family member;Other online...   
5           Books / Physical media;Online Courses or Certi...   

                                              LearnCodeOnline  \
ResponseId                                                      
1                                                         NaN   
2           Formal documentation provided by the owner of ...   
3           Formal documentation provided by the owner of ...   
4           Formal documentation provided by the owner of ...   
5           Formal documentation provided by the owner of ...   

            LearnCodeCoursesCert YearsCode YearsCodePro  \
ResponseId                                                
1                            NaN       NaN          NaN   
2                          Other        18            9   
3                            NaN        27           23   
4                            NaN        12            7   
5           Other;Codecademy;edX         6            4   

                                         DevType                   OrgSize  \
ResponseId                                                                   
1                                            NaN                       NaN   
2           Senior Executive (C-Suite, VP, etc.)          2 to 9 employees   
3                            Developer, back-end  5,000 to 9,999 employees   
4                           Developer, front-end      100 to 499 employees   
5                 

## Getting a sense of the data

There are different way to know about the data.

In [46]:
df.shape

(89184, 83)

So, we can see that this CSV has 

1. Rows: 89,184
2. Columns: 84

When we did a head on the dataframe, we can only see 20 columsn, because by default it gets trimmed.

Even with this, I am not able to view all the columns.

So, I will confifure the head to show all.

In [47]:
pd.set_option('display.max_columns', 84)
# df.set_index('ResponseId', inplace=True) # adding the index to the df
df.head()

Q120                      MainBranch              Age  \
ResponseId                                                             
1           I agree                   None of these  18-24 years old   
2           I agree  I am a developer by profession  25-34 years old   
3           I agree  I am a developer by profession  45-54 years old   
4           I agree  I am a developer by profession  25-34 years old   
5           I agree  I am a developer by profession  25-34 years old   

                                                   Employment  \
ResponseId                                                      
1                                                         NaN   
2                                         Employed, full-time   
3                                         Employed, full-time   
4                                         Employed, full-time   
5           Employed, full-time;Independent contractor, fr...   

                                      RemoteWork  \
ResponseId                                         
1                                            NaN   
2                                         Remote   
3           Hybrid (some remote, some in-person)   
4           Hybrid (some remote, some in-person)   
5                                         Remote   

                                             CodingActivities  \
ResponseId                                                      
1                                                         NaN   
2           Hobby;Contribute to open-source projects;Boots...   
3           Hobby;Professional development or self-paced l...   
4                                                       Hobby   
5           Hobby;Contribute to open-source projects;Profe...   

                                                 EdLevel  \
ResponseId                                                 
1                                                    NaN   
2           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
5           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                                    LearnCode  \
ResponseId                                                      
1                                                         NaN   
2           Books / Physical media;Colleague;Friend or fam...   
3           Books / Physical media;Colleague;On the job tr...   
4           Colleague;Friend or family member;Other online...   
5           Books / Physical media;Online Courses or Certi...   

                                              LearnCodeOnline  \
ResponseId                                                      
1                                                         NaN   
2           Formal documentation provided by the owner of ...   
3           Formal documentation provided by the owner of ...   
4           Formal documentation provided by the owner of ...   
5           Formal documentation provided by the owner of ...   

            LearnCodeCoursesCert YearsCode YearsCodePro  \
ResponseId                                                
1                            NaN       NaN          NaN   
2                          Other        18            9   
3                            NaN        27           23   
4                            NaN        12            7   
5           Other;Codecademy;edX         6            4   

                                         DevType                   OrgSize  \
ResponseId                                                                   
1                                            NaN                       NaN   
2           Senior Executive (C-Suite, VP, etc.)          2 to 9 employees   
3                            Developer, back-end  5,000 to 9,999 employees   
4                           Developer, front-end      100 to 499 employees   
5                 

Great, now I am able to view the entire data.

It's time to load the schema and see the questions.

So, I will load the other CSV file that I for in the zip file.

In [55]:
schema_df = pd.read_csv('../../data/stackoverflow/survey_results_schema.csv', index_col='qid')
schema_df.sort_index(inplace=True)
schema_df.head()

qname                                           question  \
qid                                                                      
QID1             S1  <span style="font-size:22px; font-family: aria...   
QID100  SOVisitFreq  How frequently would you say you visit Stack O...   
QID101    SOAccount              Do you have a Stack Overflow account?   
QID102   SOPartFreq  How frequently would you say you participate i...   
QID106       SOComm  Do you consider yourself a member of the Stack...   

       force_resp type selector  
qid                              
QID1        False   DB       TB  
QID100      False   MC     SAVR  
QID101      False   MC     SAVR  
QID102      False   MC     SAVR  
QID106      False   MC     SAVR

I am able to load the data. However, I can see that the question has HTML as well.

# Checking iloc and loc

In [49]:
schema_df.iloc[[1, 2, 3], 1]

qid
QID12                                     Browser Meta Info
QID310    <div><span style="font-size:19px;"><strong>You...
QID312                                                     
Name: question, dtype: object

In [50]:
schema_df.iloc[0:10, [0,1,3]]

qname                                           question  \
qid                                                                           
QID16                 S0  <div><span style="font-size:19px;"><strong>Hel...   
QID12           MetaInfo                                  Browser Meta Info   
QID310              Q310  <div><span style="font-size:19px;"><strong>You...   
QID312              Q120                                                      
QID1                  S1  <span style="font-size:22px; font-family: aria...   
QID2          MainBranch  Which of the following options best describes ...   
QID127               Age                                What is your age? *   
QID296        Employment  Which of the following best describes your cur...   
QID308        RemoteWork  Which best describes your current work situation?   
QID297  CodingActivities  Which of the following best describes the code...   

        type  
qid           
QID16     DB  
QID12   Meta  
QID310    DB  
QID312    MC  
QID1      DB  
QID2      MC  
QID127    MC  
QID296    MC  
QID308    MC  
QID297    MC

Looking at the iloc feature of DataFrames.

In [51]:
df.loc[0:4, 'Age':'EdLevel']

Age           Employment  \
ResponseId                                         
1           18-24 years old                  NaN   
2           25-34 years old  Employed, full-time   
3           45-54 years old  Employed, full-time   
4           25-34 years old  Employed, full-time   

                                      RemoteWork  \
ResponseId                                         
1                                            NaN   
2                                         Remote   
3           Hybrid (some remote, some in-person)   
4           Hybrid (some remote, some in-person)   

                                             CodingActivities  \
ResponseId                                                      
1                                                         NaN   
2           Hobby;Contribute to open-source projects;Boots...   
3           Hobby;Professional development or self-paced l...   
4                                                       Hobby   

                                                 EdLevel  
ResponseId                                                
1                                                    NaN  
2           Bachelor’s degree (B.A., B.S., B.Eng., etc.)  
3           Bachelor’s degree (B.A., B.S., B.Eng., etc.)  
4           Bachelor’s degree (B.A., B.S., B.Eng., etc.)

# Analysing data

In [56]:
df.head()

Q120                      MainBranch              Age  \
ResponseId                                                             
1           I agree                   None of these  18-24 years old   
2           I agree  I am a developer by profession  25-34 years old   
3           I agree  I am a developer by profession  45-54 years old   
4           I agree  I am a developer by profession  25-34 years old   
5           I agree  I am a developer by profession  25-34 years old   

                                                   Employment  \
ResponseId                                                      
1                                                         NaN   
2                                         Employed, full-time   
3                                         Employed, full-time   
4                                         Employed, full-time   
5           Employed, full-time;Independent contractor, fr...   

                                      RemoteWork  \
ResponseId                                         
1                                            NaN   
2                                         Remote   
3           Hybrid (some remote, some in-person)   
4           Hybrid (some remote, some in-person)   
5                                         Remote   

                                             CodingActivities  \
ResponseId                                                      
1                                                         NaN   
2           Hobby;Contribute to open-source projects;Boots...   
3           Hobby;Professional development or self-paced l...   
4                                                       Hobby   
5           Hobby;Contribute to open-source projects;Profe...   

                                                 EdLevel  \
ResponseId                                                 
1                                                    NaN   
2           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
5           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                                    LearnCode  \
ResponseId                                                      
1                                                         NaN   
2           Books / Physical media;Colleague;Friend or fam...   
3           Books / Physical media;Colleague;On the job tr...   
4           Colleague;Friend or family member;Other online...   
5           Books / Physical media;Online Courses or Certi...   

                                              LearnCodeOnline  \
ResponseId                                                      
1                                                         NaN   
2           Formal documentation provided by the owner of ...   
3           Formal documentation provided by the owner of ...   
4           Formal documentation provided by the owner of ...   
5           Formal documentation provided by the owner of ...   

            LearnCodeCoursesCert YearsCode YearsCodePro  \
ResponseId                                                
1                            NaN       NaN          NaN   
2                          Other        18            9   
3                            NaN        27           23   
4                            NaN        12            7   
5           Other;Codecademy;edX         6            4   

                                         DevType                   OrgSize  \
ResponseId                                                                   
1                                            NaN                       NaN   
2           Senior Executive (C-Suite, VP, etc.)          2 to 9 employees   
3                            Developer, back-end  5,000 to 9,999 employees   
4                           Developer, front-end      100 to 499 employees   
5                 

In [52]:
df['RemoteWork'].value_counts()

RemoteWork
Hybrid (some remote, some in-person)    31131
Remote                                  30566
In-person                               12113
Name: count, dtype: int64

Using this, we are able to understand that a significant percentage of users have opted for Hybrid.

In person is very low compared to other two options.

In [67]:
def clean_years_of_code(value):
    try:
        return float(value)
    except ValueError:
        return None

In [68]:
df['YearsCode'] = df['YearsCode'].apply(clean_years_of_code)
df['YearsCode'].unique()

array([nan, 18., 27., 12.,  6., 21.,  4.,  5., 20., 14., 10., 15., 11.,
        3., 24.,  8., 13., 16., 33., 22., 30., 32.,  7., 35., 28., 40.,
       17., 29., 19.,  9., 38., 26., 34., 25.,  2., 45., 23., 31., 43.,
        1., 48., 41., 50., 39., 42., 37., 36., 44., 46., 49., 47.])

In [69]:
df['YearsCode'].value_counts()

YearsCode
10.0    6521
5.0     5415
6.0     4893
8.0     4879
7.0     4800
4.0     4466
15.0    4336
3.0     4269
20.0    4021
12.0    3471
9.0     3130
2.0     2826
25.0    2685
13.0    2601
11.0    2519
14.0    2278
30.0    2078
16.0    1842
18.0    1690
17.0    1544
23.0    1327
40.0    1302
1.0     1294
22.0    1210
35.0    1118
24.0     959
19.0     826
21.0     761
28.0     715
27.0     710
26.0     706
33.0     496
32.0     444
38.0     428
45.0     360
42.0     347
34.0     333
29.0     324
37.0     312
36.0     282
31.0     272
43.0     265
39.0     229
41.0     227
44.0     160
50.0     125
46.0     101
48.0      96
47.0      79
49.0      42
Name: count, dtype: int64